### PR TITLE
Descriptor systems for humans

### PR DESCRIPTION
**Pull request by Azarak**
## About The Pull Request
This adds a system which allows to loosely attach descriptor types to humans, which are used to describe a property of him, and they can be used for comparisons too.
Currently there's only ones for age, age of robots, and strength.
dreamseeker_2022-05-08_22-01-36
dreamseeker_2022-05-08_22-02-23
(The improper "looks" instead of varaible "look/s" is fixed now)

In another PR I'll add height and maybe weight too because adding those features in itself will be a bit bigger scope due to needing all sorts of species based customization and prefs validations etc.

## Changelog
:cl:
add: Descriptor systems for humans
/:cl:

### CODE REVIEW

File: code/modules/mob/living/carbon/carbon_defines.dm [hunk 1]
```diff
@@ -107,3 +107,8 @@
 
 	/// Cooldown for the next smell
 	var/next_smell = 0
+
+	/// List of descriptors this human has. Yes this has to be on a carbon level to be compatible with species
+	var/list/descriptors = list(
+		/datum/descriptor/attribute/strength
+		)
```
Comment (atakiya) [inline code/modules/mob/living/carbon/carbon_defines.dm:114]:
```suggestion
	)
```

File: code/modules/mob/living/carbon/human/descriptor/descriptors.dm [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+	return SPAN_SMALLNOTICE("[examined.p_they(TRUE)] appear[examined.p_s()] to be [compare_text]")
+
+/// Descriptor of age for robots
+/datum/descriptor/age/robot
+	
+/datum/descriptor/age/robot/describe(mob/living/carbon/human/examined, description_value)
+	var/age_text
+	switch(description_value)
+		if(1)
+			age_text = "fresh out of factory."
```
Comment (atakiya) [inline code/modules/mob/living/carbon/human/descriptor/descriptors.dm:73]:
```suggestion
			age_text = "fresh out of the factory."
```
Alternatively, `fresh from the factory`?

File: code/modules/mob/living/carbon/human/descriptor/descriptors.dm [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+	var/raw = TRUE
+
+/datum/descriptor/attribute/description_value(mob/living/carbon/human/examined)
+	var/attribute
+	if(raw)
+		attribute = examined.attributes.attributes_raw[attribute_type]
+	else
+		attribute = examined.attributes.attributes_final[attribute_type]
+	// Will return a 1 or -1 per each 2 attributes that move away from the base value. So 12 is 1, 8 is -1 etc.
+	return FLOOR((attribute - ATTRIBUTE_BASE) / 2, 1)
```
Comment (atakiya) [inline code/modules/mob/living/carbon/human/descriptor/descriptors.dm:121]:
```suggestion
	return FLOOR((attribute - ATTRIBUTE_BASE) * 0.5, 1)
```

File: code/modules/mob/living/carbon/human/descriptor/descriptors.dm [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+
+/datum/descriptor/attribute/strength/describe(mob/living/carbon/human/examined, description_value)
+	var/describe_text
+	switch(description_value)
+		if(-INFINITY to -2)
+			describe_text = "look[examined.p_s()] very weak."
+		if(-1)
+			describe_text = "look[examined.p_s()] weak."
+		if(0)
+			describe_text = "[examined.p_have()] moderate musculature."
```
Comment (atakiya) [inline code/modules/mob/living/carbon/human/descriptor/descriptors.dm:139]:
```suggestion
			describe_text = "[examined.p_s()] average."
```
"[has] moderate musculature" seems a bit out of place with the rest of the strings in this category.

File: code/modules/mob/living/carbon/human/descriptor/human_descriptor.dm [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+		if(compare)
+			// Human type on user is asserted here thanks to the first check in the proc.
+			string += "\t<i>[descriptor.get_comparison(user, src)]</i>"
+		else
+			string += "\t<i>[descriptor.get_description(src)]</i>"
+		first = FALSE
+
+	/// If a human user is looking at a normal description, add a link to do a comparison too.
+	if(is_user_human && user != src && !compare && has_any_readable_descriptors(TRUE, user))
+		string += "<BR>\t<a href='?src=[REF(src)];lookup_info=compare'>Compare yourself to [src.p_them()].</a>"
```
Comment (atakiya) [inline code/modules/mob/living/carbon/human/descriptor/human_descriptor.dm:25]:
```suggestion
		string += "<br/>\t<a href='?src=[REF(src)];lookup_info=compare'>Compare yourself to [src.p_them()].</a>"
```
If all the other tags are lowercase, lets do this one too. Bonus points for autoclosing.

File: code/modules/mob/living/carbon/human/species.dm [hunk 6]
```diff
@@ -274,6 +274,10 @@ GLOBAL_LIST_EMPTY(customizable_races)
 			'sound/voice/scream_f2.ogg',
 		)
 	)
+	/// List of descriptors related to this species
+	var/list/species_descriptors = list(
+		/datum/descriptor/age
+		)
```
Comment (atakiya) [inline code/modules/mob/living/carbon/human/species.dm:280]:
```suggestion
	)
```

File: code/modules/mob/living/carbon/human/species_types/robotic.dm [hunk 7]
```diff
@@ -37,6 +37,9 @@
 	scream_sounds = list(
 		NEUTER = 'sound/voice/scream_silicon.ogg'
 	)
+	species_descriptors = list(
+		/datum/descriptor/age/robot
+		)
```
Comment (atakiya) [inline code/modules/mob/living/carbon/human/species_types/robotic.dm:42]:
```suggestion
	)
```

File: code/modules/mob/living/carbon/human/descriptor/descriptors.dm (hunk 4 above)
Comment (Azarak) [reply code/modules/mob/living/carbon/human/descriptor/descriptors.dm:139]: But if you just say "average" you don't specify what it is.

Comment (atakiya) [reply code/modules/mob/living/carbon/human/descriptor/descriptors.dm:139]: Hmm..

Comment (atakiya) [reply code/modules/mob/living/carbon/human/descriptor/descriptors.dm:139]:
```suggestion
			describe_text = "look[examined.p_s()] average in strength."
```

Comment (Azarak) [reply code/modules/mob/living/carbon/human/descriptor/descriptors.dm:139]: yeah seems fine

**Review by atakiya — APPROVED** [review]

**Pull request merged**